exercises/circular-buffer/example.jl

```diff
@@ -0,0 +1,177 @@
+module CircularBuffers
+
+using Base: @propagate_inbounds
+using Printf
+
+export CircularBuffer, capacity, isfull
+
+const OVERWRITE = false
+
+"""
+    CircularBuffer{T} <: AbstractVector{T}
+
+Circular buffer with elements of type `T`, useful as a FIFO or LIFO queue with fixed
+capacity.
+
+Optimized for inserting elements at the end¸ so `push!` performs better (~1.5x faster)
+than `pushfirst!`. Deleting elements at either end (`pop!`, and `popfirst!`) as well as
+`getindex` and `setindex!` on random elements are all efficient.
+
+---
+
+    CircularBuffer(v::Vector{T}) -> CircularBuffer{T}
+
+Construct a `CircularBuffer{T}` using `v` as the backing data without copying. Its capacity
+and length are equal to `length(v)`.
```

Comment:
Good point. I looked around Julia Base and it looks like the norm there at least is to organize things to avoid documenting both the type and an inner constructor at the same time. In any case, I've now massively simplified the constructors into a single inner constructor, since the rest are unnecessary for the exercise as you pointed out.